File: algorithms/easy/108_convert_sorted_array_to_binary_search_tree.py
# 108. Convert Sorted Array to Binary Search Tree

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def sorted_array_to_bst(nums):
    """Convert a sorted array to a height-balanced binary search tree.
    
    A height-balanced binary tree is a binary tree in which the depth of the 
    two subtrees of every node never differs by more than one.
    
    Args:
        nums: Sorted array of integers
    
    Returns:
        Root of the balanced BST
    """
    if not nums:
        return None
    
    # Find the middle element to make it the root
    mid = len(nums) // 2
    
    # Create the root node with the middle element
    root = TreeNode(nums[mid])
    
    # Recursively build the left subtree with elements before the middle
    root.left = sorted_array_to_bst(nums[:mid])
    
    # Recursively build the right subtree with elements after the middle
    root.right = sorted_array_to_bst(nums[mid+1:])
    
    return root

# More efficient implementation using indices instead of slicing
def sorted_array_to_bst_optimized(nums):
    def helper(left, right):
        if left > right:
            return None
        
        # Find the middle element
        mid = (left + right) // 2
        
        # Create the root node with the middle element
        root = TreeNode(nums[mid])
        
        # Recursively build the left and right subtrees
        root.left = helper(left, mid - 1)
        root.right = helper(mid + 1, right)
        
        return root
    
    return helper(0, len(nums) - 1)

# Helper function to perform inorder traversal for testing
def inorder_traversal(root):
    result = []
    
    def inorder(node):
        if not node:
            return
        inorder(node.left)
        result.append(node.val)
        inorder(node.right)
    
    inorder(root)
    return result

# Test cases
# Example 1: [-10,-3,0,5,9] => [0,-3,9,-10,null,5]
root1 = sorted_array_to_bst([-10, -3, 0, 5, 9])
print(inorder_traversal(root1))  # [-10, -3, 0, 5, 9] (should match original array)

# Example 2: [1,3] => [3,1]
root2 = sorted_array_to_bst([1, 3])
print(inorder_traversal(root2))  # [1, 3] (should match original array)

File: algorithms/easy/108_convert_sorted_array_to_binary_search_tree.js
/**
 * 108. Convert Sorted Array to Binary Search Tree
 * 
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function(nums) {
    if (nums.length === 0) return null;
    
    // Find the middle element to make it the root
    const mid = Math.floor(nums.length / 2);
    
    // Create the root node with the middle element
    const root = new TreeNode(nums[mid]);
    
    // Recursively build the left subtree with elements before the middle
    root.left = sortedArrayToBST(nums.slice(0, mid));
    
    // Recursively build the right subtree with elements after the middle
    root.right = sortedArrayToBST(nums.slice(mid + 1));
    
    return root;
};

// More efficient implementation using indices instead of slicing
var sortedArrayToBSTOptimized = function(nums) {
    const helper = function(left, right) {
        if (left > right) return null;
        
        // Find the middle element
        const mid = Math.floor((left + right) / 2);
        
        // Create the root node with the middle element
        const root = new TreeNode(nums[mid]);
        
        // Recursively build the left and right subtrees
        root.left = helper(left, mid - 1);
        root.right = helper(mid + 1, right);
        
        return root;
    };
    
    return helper(0, nums.length - 1);
};

// Helper function to perform inorder traversal for testing
function inorderTraversal(root) {
    const result = [];
    
    function inorder(node) {
        if (!node) return;
        inorder(node.left);
        result.push(node.val);
        inorder(node.right);
    }
    
    inorder(root);
    return result;
}

// Test cases
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

// Example 1: [-10,-3,0,5,9] => [0,-3,9,-10,null,5]
const root1 = sortedArrayToBST([-10, -3, 0, 5, 9]);
console.log(inorderTraversal(root1));  // [-10, -3, 0, 5, 9] (should match original array)

// Example 2: [1,3] => [3,1]
const root2 = sortedArrayToBST([1, 3]);
console.log(inorderTraversal(root2));  // [1, 3] (should match original array)